### DISCUSSION TITLE
Question about 16 bit colour formats

### DISCUSSION THREAD

**Discussion by QuietClaws** [MonoGame/MonoGame#8546] [Graphics]
For a project I'm working on I was thinking about using 16 bit colour instead of 32 bit colour for textures so I could store more sprite frames in the same amount of memory.  I looked into compression, but I didn't like the artifacts it leaves around the sprites where the transparency is because of how it combines multiple pixels.

I was experimenting with some formats and monitoring the video memory usage.  

I saved some PNGs as 16 bit (RGBA4444).  I found in Content Manager if I switch from Color to Color16bit it seems to use the same amount of memory.  I'm assuming "16 bit" means total, not per element.  I also tried "no change" format as well, but it didn't seem to help.

I tried creating Texture2Ds manually with SurfaceFormat.Bgra4444 and the memory usage seemed the same as SurfaceFormat.Color.  However if I manually create a Bgra5551 the memory use is smaller.  

I was thinking that Bgra4444, Color16 in Content Manager and Bgra5551 would use the same video memory - 16 bits per pixel and this would be less than the use for Color which would be 32 bpp.

If anyone has any insights on using lower numbers of colours to save memory I'd appreciate it.

**Comment by ThomasFOG** [comment]
First things first: what tool are you using to measure GPU VRAM usage? Because there's actually no tool that will give you a correct measure.

GPUs are very opaque things that don't quite give you what you think when it comes to numbers. Sometime they allocate stuff ahead and it may appear that what should use a precise amount of memory seems to be using more than it should. This is vendor/driver specific and not accessible by systems, so there is technically no software capable of knowing per-process GPU VRAM usage precisely.

If we want to check exactly if this works as expected, we should check what goes into ```Texture2D.PlatformSetData()``` when the texture is constructed. If the data/elemCount are of the expected size, then that's what goes to the GPU even though software measures tell otherwise.

(Maybe we could add some metrics to MonoGame to know what's currently allocated on the GPU.)

**Reply by QuietClaws** [reply]
Thanks very much for your reply.  It's a good point about the graphics card and that may be the answer.

I wasn't looking for exact memory usage, but allocated quite a few textures to see overall by monitoring in GPU-Z.  Windows Task Manager had similar numbers.

Just to give a simple example:

Texture2D[] setOfTextures = new Texture2D[100];

In LoadContent(), if I allocate a bunch of textures like this:
	int q;
    for (q = 0; q < 100; q++)
        setOfTextures[q] = new Texture2D(_graphics.GraphicsDevice, 4096, 4096, false, SurfaceFormat.Bgra4444); 
	
Then in Draw(), something like this (have to use it or it's smart enough not to allocate the texture memory):
	for (q = 0; q < 100; q++)
		_spriteBatch.Draw(setOfTextures[q], new Vector2(q * 10.0f, 0), Color.White);

I haven't filled the textures, so I guess they won't have anything to draw, but it still seems to use VRAM.  

I have an RTX 3070 with 8GB Ram.  If I'm doing "virtually nothing" on windows it's using around 900 MB dedicated memory/VRAM use.  
If I use SurfaceFormat.Bgra5551 in the above new Texture, then GPU-Z is showing around 4000 MB.  This seems reasonable with screen buffers, other items, etc. [32MB x 100 textures = ~3.2 GB]
If I change to SurfaceFormat.Bgra4444, then GPU-Z is showing around 7200 MB used (much more than Bgra5551).  
If I change to SurfaceFormat.Color, then GPU-Z is showing around 7200 MB used (around the same as Bgra4444)

If I save TGAs and set 16 bit colour in content manager, running with a TGA exported as 5551 uses much less vram than a TGA exported with 4444; the 4444 one will use use same/similar memory to 32 bpp colour.

As you said, maybe it's something specific to the graphics card implementation, because if I extract the data from a 4444 texture with the expected 16 bit (2 byte size) it doesn't give an error.
            byte[] c = new byte[4096 * 4096 * 2];
            setOfTextures[0].GetData<byte>(c);

It's also possible I'm doing something wrong.  It's not a big deal; I can work around it.  I was just curious if anyone else had any experience.

**Reply by ThomasFOG** [reply]
GPU-Z and the Windows Task Manager both sure are unreliable to precisely measure VRAM usage, and the behavior that you describe can be a lot of different things like the system/driver caching stuff or the GPU allocating stuff ahead for the context. The differences depending on surface format could very well be because the GPU or the system/driver have a different behavior depending on formats (like optimizations, or being forced to arrange memory in a specific way which ends up using more VRAM than it should; this is very likely vendor, and per-chip specific, and not documented by vendors).

> I haven't filled the textures, so I guess they won't have anything to draw, but it still seems to use VRAM.

That's expected. When you create an empty ```Texture2D```, MonoGame will create the GPU buffer for that texture to reserve the memory (it basically creates and uploads a blank texture on the GPU). In the case of a 4096x4096 ```Bgra4444``` texture, it will allocate 32MB of VRAM (which is your ```byte[4096 * 4096 *2]``` later on).

What the GPU will do is unfortunately a black box, and you can't take for granted that one specific format seems to take less VRAM is a reproductible behavior across vendors, drivers, or chips of the same vendors.

What's certain, is that MonoGame doesn't convert formats before uploading them to the GPU and what you provide to ```SetData()``` is what is sent to the driver for upload to the GPU. What happens after is a black box.
